src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Permissions/DefaultPermissionService.cs

```diff
@@ -0,0 +1,58 @@
+namespace OrchardCore.Security.Permissions;
+
+public sealed class DefaultPermissionService : IPermissionService
+{
+    private readonly IEnumerable<IPermissionProvider> _permissionProviders;
+
+    // Cached per request.
+    private Dictionary<string, Permission> _permissions;
+
+    public DefaultPermissionService(IEnumerable<IPermissionProvider> permissionProviders)
+    {
+        _permissionProviders = permissionProviders;
+    }
+
+    public async ValueTask<Permission> FindByNameAsync(string name)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(name);
+
+        if (_permissions is null)
+        {
+            await LoadPermissionsAsync();
+        }
+
+        if (_permissions.TryGetValue(name, out var permission))
+        {
+            return permission;
+        }
+
+        return null;
+    }
+
+    public async ValueTask<IEnumerable<Permission>> GetPermissionsAsync()
+    {
+        if (_permissions != null)
+        {
+            return _permissions.Values;
+        }
+
+        await LoadPermissionsAsync();
+
+        return _permissions.Values;
+    }
+
+    private async Task LoadPermissionsAsync()
```

Comment:
Not sure there is a reason to create this. GetPermissionsAsync could do the same directly. Even if you don't need the result.

Reply:
Agreed. But I find it more readable this way.